orchagent/srv6orch.cpp

```diff
@@ -89,6 +95,222 @@ static bool mySidDscpModeToSai(const string& mode, sai_tunnel_dscp_mode_t& sai_m
+{
+    auto mysid_addr = getMySidAddress(sai_entry).to_string();
+    auto locator_cfg = getMySidEntryLocatorCfg(sai_entry);
+    return getMySidPrefix(mysid_addr, locator_cfg);
+}
+
+bool Srv6Orch::addMySidCounter(const sai_my_sid_entry_t& sai_entry, sai_object_id_t& counter_oid)
+{
+    SWSS_LOG_ENTER();
+
+    if (!FlowCounterHandler::createGenericCounter(counter_oid))
+    {
+        SWSS_LOG_ERROR("Failed to create SAI counter for SRv6 MySID entry");
+        return false;
+    }
+
+    auto key = getMySidCounterKey(sai_entry);
+    vector<FieldValueTuple> fvs = {
+        {key, sai_serialize_object_id(counter_oid)}
+    };
+
+    m_mysid_counters_table->set("", fvs);
+
+    auto was_empty = m_pending_counters.empty();
+    m_pending_counters[counter_oid] = key;
+
+    if (was_empty)
+    {
+        m_counter_update_timer->start();
+    }
+
+    return true;
+}
+
+void Srv6Orch::removeMySidCounter(const sai_my_sid_entry_t& sai_entry, sai_object_id_t& counter_oid)
+{
+    SWSS_LOG_ENTER();
+
+    if (counter_oid == SAI_NULL_OBJECT_ID)
+    {
+        return;
+    }
+
+    auto key = getMySidCounterKey(sai_entry);
+
+    m_mysid_counters_table->hdel("", key);
+
+    auto was_pending = m_pending_counters.erase(counter_oid) == 1;
+    if (!was_pending)
+    {
+        SWSS_LOG_INFO("Unregistering SRv6 counter for %s, oid %s", key.c_str(), sai_serialize_object_id(counter_oid).c_str());
+        m_counter_manager.clearCounterIdList(counter_oid);
+    }
+
+    FlowCounterHandler::removeGenericCounter(counter_oid);
+    counter_oid = SAI_NULL_OBJECT_ID;
+}
+
+void Srv6Orch::setMySidEntryCounter(const sai_my_sid_entry_t& sai_entry, sai_object_id_t counter_oid)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_MY_SID_ENTRY_ATTR_COUNTER_ID;
+    attr.value.oid = counter_oid;
+
+    auto status = sai_srv6_api->set_my_sid_entry_attribute(&sai_entry, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set my_sid entry counter oid to %s, rc: %s", sai_serialize_object_id(counter_oid).c_str(), sai_serialize_status(status).c_str());
+    }
+}
+
+void Srv6Orch::setCountersState(bool enable)
+{
+    SWSS_LOG_ENTER();
+
+    if (!getMySidCountersSupported())
+    {
+        SWSS_LOG_WARN("Ignoring SRv6 counters state change as they are not supported on this platform");
+        return;
+    }
+
+    if (enable == m_mysid_counters_enabled)
+    {
+        return;
+    }
+
+    SWSS_LOG_INFO("Setting SRv6 MySID counters state to %s", enable ? "enabled" : "disabled");
+
+    for (auto& mysid : srv6_my_sid_table_)
+    {
+        const auto& sai_entry = mysid.second.entry;
+        auto &counter_oid = mysid.second.counter;
+
+        if (enable)
+        {
+            addMySidCounter(sai_entry, counter_oid);
+            setMySidEntryCounter(sai_entry, counter_oid);
+        } else {
+            setMySidEntryCounter(sai_entry, SAI_NULL_OBJECT_ID);
```

Comment:
Will it impact counter-polling? If sairedis happens to poll the counter that is deregistered from MySid but has not yet stopped polling.